### PR TITLE
Cosmos: Stop generating discriminator clause when not needed 

### LINKED ISSUE #20268: Cosmos: Leverage complete discriminator mapping
* Issue more efficient queries
* Validate all types mapped to a collection have the same configuration

### PR DESCRIPTION
**Pull request by roji**
**Note: this PR is based on top of #34143, review last commit only**

@AndriySvyryd this doesn't do validation (which is also part of #20268), we can split that out or if if you feel like it, just push it into this PR?

Closes #20268

### CODE REVIEW

File: src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs [hunk 1]
```diff
@@ -427,37 +427,33 @@ protected override ShapedQueryExpression CreateShapedQueryExpression(IEntityType
 
         // Add discriminator predicate
         var concreteEntityTypes = entityType.GetConcreteDerivedTypesInclusive().ToList();
-        if (concreteEntityTypes.Count == 1)
+        if (concreteEntityTypes is [var singleEntityType]
+            && singleEntityType.GetIsDiscriminatorMappingComplete()
+            && entityType.GetContainer() is var container
+            && !entityType.Model.GetEntityTypes().Any(e => e.GetContainer() == container && e != singleEntityType))
```
Comment (roji) [inline src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:433]: @AndriySvyryd does this seem right? When do we get a relational model for Cosmos :trollface:

File: src/EFCore.Cosmos/Query/Internal/SqlExpressionFactory.cs [hunk 2]
```diff
@@ -609,16 +609,18 @@ public virtual SqlConditionalExpression Condition(SqlExpression test, SqlExpress
     ///     any release. You should only use it directly in your code with extreme caution and knowing that
     ///     doing so can result in application failures when updating to a new Entity Framework Core release.
     /// </summary>
-    public virtual InExpression In(SqlExpression item, IReadOnlyList<SqlExpression> values)
-        => ApplyTypeMappingOnIn(new InExpression(item, values, _boolTypeMapping));
+    public virtual SqlExpression In(SqlExpression item, IReadOnlyList<SqlExpression> values)
+        => values is [var singleValue]
```
Comment (roji) [inline src/EFCore.Cosmos/Query/Internal/SqlExpressionFactory.cs:613]: @ranma42 continuing with the trend of factory basic optimizations...

**Comment by AndriySvyryd** [comment]
> @AndriySvyryd this doesn't do validation (which is also part of https://github.com/dotnet/efcore/issues/20268), we can split that out or if you feel like it, just push it into this PR?

The validation should be pretty simple, just add it

**Comment by roji** [comment]
> Validate all types mapped to a collection have the same configuration

Which configuration exactly?

File: src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs (hunk 1 above)
Comment (roji) [reply src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:433]: Do we support that? What would that query look like, given that there's no common root?

**Comment by AndriySvyryd** [comment]
> Which configuration exactly?

`IsDiscriminatorMappingComplete`. It doesn't make sense if it's not set to the same value on all entity types sharing the container.

File: src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:433]: I am specifically talking about the case where there's a common root

**Comment by roji** [comment]
@AndriySvyryd done

File: src/EFCore.Cosmos/Properties/CosmosStrings.resx [hunk 3]
```diff
@@ -162,6 +162,9 @@
   <data name="InvalidResourceId" xml:space="preserve">
     <value>Unable to generate a valid 'id' value to execute a 'ReadItem' query. This usually happens when the value provided for one of the properties is 'null' or an empty string. Please supply a value that's not 'null' or an empty string.</value>
   </data>
+  <data name="IsDiscriminatorMappingCompleteMismatch" xml:space="preserve">
+    <value>The IsDiscriminatorMappingComplete setting was configured to '{isDiscriminatorMappingComplete1}' on '{entityType1}', but on '{entityType2}' it was configured to '{isDiscriminatorMappingComplete2}'. All entity types mapped to the same container '{container}' must be configured with the same IsDiscriminatorMappingComplete settings.</value>
```
Comment (AndriySvyryd) [inline src/EFCore.Cosmos/Properties/CosmosStrings.resx:166]:
```suggestion
    <value>The IsDiscriminatorMappingComplete setting was configured to '{isDiscriminatorMappingComplete1}' on '{entityType1}', but on '{entityType2}' it was configured to '{isDiscriminatorMappingComplete2}'. All entity types mapped to the same container '{container}' must be configured with the same IsDiscriminatorMappingComplete value.</value>
```

File: src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:433]: And this part can be done in a separate PR, as it's less impactful.

**Review by AndriySvyryd — APPROVED** [review]

File: src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs (hunk 1 above)
Comment (roji) [reply src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:433]: Let's do that separately, ok? I don't want to spend the time doing a test suite, as this doesn't seem to be a covered scenario in any case.

**Comment by roji** [comment]
Thanks for the quick review @AndriySvyryd, will need to merge the base PRs first before I can do this one.

File: src/EFCore.Cosmos/Query/Internal/SqlExpressionFactory.cs (hunk 2 above)
Comment (ranma42) [reply src/EFCore.Cosmos/Query/Internal/SqlExpressionFactory.cs:613]: iiuc this also means that `x IN (y)` has the same semantics as `x == y` along the pipeline, right?
(aka it changes semantics in the `SqlNullabilityProcessor`)

Comment (roji) [reply src/EFCore.Cosmos/Query/Internal/SqlExpressionFactory.cs:613]: Yes - AFAIK in SQL `x IN (a, b, c)` is identical to `x = a OR x = b OR x = c` in all respects (including null semantics). And yeah, the semantics change from 2-value to 3-value logic in SqlNullabilityProcessor, just as they do for equality.

Comment (ranma42) [reply src/EFCore.Cosmos/Query/Internal/SqlExpressionFactory.cs:613]: Yep, the important part here is that the pipeline before `SqlNullabilityProcessor` should consider `null IN (1, null)` as evaluating to true, while the pipeline after `SqlNullabilityProcessor` should consider it as evaluating to `null`.

Comment (roji) [reply src/EFCore.Cosmos/Query/Internal/SqlExpressionFactory.cs:613]: Exactly.

**Pull request merged**